gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/FSJobCatalog.java

```diff
@@ -0,0 +1,288 @@
+   * return null requested URI not existed
+   * requires null checking
+   * @param uri The relative Path to the target job configuration.
+   * @return
+   */
+  @Override
+  public synchronized JobSpec getJobSpec(URI uri) {
+    Path targetJobSpecPath = new Path(uri);
+
+    List<JobSpec> resultJobSpecList =
+        FSJobCatalogHelper.loadJobConfigHelper(this.jobConfDirPath, Action.SINGLEJOB, Optional.of(targetJobSpecPath));
+    if (resultJobSpecList == null || resultJobSpecList.size() == 0) {
+      LOGGER.warn("No JobSpec with URI:" + uri + " is found.");
+      return null;
+    } else {
+      return resultJobSpecList.get(0);
+    }
+  }
+
+  /**
+   * For each new coming JobCatalogListener, react accordingly to add all existing JobSpec.
+   * @param jobListener
+   */
+  @Override
+  public synchronized void addListener(JobCatalogListener jobListener) {
+    Preconditions.checkNotNull(jobListener);
+
+    this.listeners.addListener(jobListener);
+
+    List<JobSpec> currentJobSpecList = this.getJobs();
+    if (currentJobSpecList == null || currentJobSpecList.size() == 0) {
+      return;
+    } else {
+      for (JobSpec jobSpecEntry : currentJobSpecList) {
+        AddJobCallback addJobCallback = new AddJobCallback(jobSpecEntry);
+        this.listeners.callbackOneListener(addJobCallback, jobListener);
+      }
+    }
+  }
+
+  @Override
+  public synchronized void removeListener(JobCatalogListener jobListener) {
+    this.listeners.removeListener(jobListener);
+  }
+
+  /**
+   * Allow user to programmatically add a new JobSpec.
+   * The method will materialized the jobSpec into real file.
+   *
+   * @param jobSpec The target JobSpec Object to be materialized.
+   *                Noted that the URI return by getUri is a relative path.
+   */
+  @Override
+  public synchronized void put(JobSpec jobSpec) {
+    Preconditions.checkNotNull(jobSpec);
+    try {
+      String tmpStr = jobSpec.getUri().toString();
+      Path tmpPath = new Path(this.jobConfDirPath, tmpStr);
+      if (fs.exists(tmpPath)) {
+        LOGGER.info("The job with URI[" + new Path(this.jobConfDirPath, jobSpec.getUri().toString())
+            + "] has been added before, will cover the original one.");
+        JobSpec oldSpec =  FSJobCatalogHelper.loadJobConfigHelper(FSJobCatalog.this.jobConfDirPath, Action.SINGLEJOB,
+            Optional.of(new Path(tmpStr))).get(0);
+        fs.delete(new Path(this.jobConfDirPath, jobSpec.getUri().toString()), false);
+        LOGGER.info("The old job deleted.");
+
+        // The checkAndNotify method cannot file update's implementation using deletion first creation followed,
+        // since each creation will invoke refresh() method that update the modification time.
+        // Here notify the gobblin Instance driver directly instead of letting the detector to notify.
+        this.listeners.onUpdateJob(oldSpec, jobSpec);
+      }
+
+      FSJobCatalogHelper.materializedJobSpec(this.jobConfDirPath, jobSpec);
```

Comment:
Generally better to first write the file in a staging location, then swap. (The write might fail, and at this point you already lost the old file).